libcore/FileUtils.vala

```diff
@@ -562,6 +562,144 @@ namespace PF.FileUtils {
                 (uri.contains (GLib.Path.DIR_SEPARATOR_S + ".local") &&
                  uri.contains (GLib.Path.DIR_SEPARATOR_S + "Trash" + GLib.Path.DIR_SEPARATOR_S)));
     }
+
+    public Gdk.DragAction file_accepts_drop (GOF.File dest,
+                                             GLib.List<GLib.File> drop_file_list, // read-only
+                                             Gdk.DragContext context,
+                                             out Gdk.DragAction suggested_action_return) {
+
+        var actions = context.get_actions ();
+        var suggested_action = context.get_suggested_action ();
+        var target_location = dest.get_target_location ();
+
+        suggested_action_return = Gdk.DragAction.PRIVATE;
+
+        if (drop_file_list == null || drop_file_list.data == null)  {
+            return Gdk.DragAction.DEFAULT;
+        }
+
+        if (dest.is_folder ()) {
+            if (!dest.is_writable ()) {
+                actions = Gdk.DragAction.DEFAULT;
+            } else {
+                /* Modify actions and suggested_action according to source files */
+                actions &= valid_actions_for_file_list (target_location,
+                                                        drop_file_list,
+                                                        ref suggested_action);
+            }
+        } else if (dest.is_executable ()) {
+            actions |= (Gdk.DragAction.COPY |
+                       Gdk.DragAction.MOVE |
+                       Gdk.DragAction.LINK |
+                       Gdk.DragAction.PRIVATE);
+        } else {
+            actions = Gdk.DragAction.DEFAULT;
+        }
+
+        if (actions == Gdk.DragAction.DEFAULT) { // No point asking if no other valid actions
+            return Gdk.DragAction.DEFAULT;
+        } else if (location_is_in_trash (target_location)) { // cannot copy or link to trash
+            actions &= ~(Gdk.DragAction.COPY | Gdk.DragAction.LINK);
+        }
+
+        if ((suggested_action & actions) != 0) {
```

Comment:
Can we use the native Vala `in` syntax here?
```if (suggested_action in actions) { ... }```

Reply:
Thanks - I did not know that was valid for enums!